pkg/validation/validator.go

```diff
@@ -0,0 +1,42 @@
+package validation
+
+import (
+	"fmt"
+	"regexp"
+)
+
+var (
+	addressValidationRegexp = regexp.MustCompile(`^(one[a-zA-Z0-9]{39})|(0x[a-fA-F0-9]{40})`)
+)
+
+// ValidateAddress validates that an address is a valid bech32 address (one...) or a valid base16 address (0x...)
+func ValidateAddress(address string) error {
+	matches := addressValidationRegexp.FindAllStringSubmatch(address, -1)
+	if len(matches) == 0 {
+		return fmt.Errorf("The address you supplied (%s) is in an invalid format. Please provide a valid address.", address)
+	}
+
+	return nil
+}
+
+// ValidateShardIDs validates senderShard and receiverShard against the shardCount
+func ValidShardIDs(senderShard int, receiverShard int, shardCount int) error {
+	if !ValidShardID(senderShard, shardCount) {
+		return fmt.Errorf("invalid argument \"%d\" for \"--from-shard\" flag: please specify a valid shard ID using --from-shard and try again!", senderShard)
+	}
+
+	if !ValidShardID(receiverShard, shardCount) {
+		return fmt.Errorf("invalid argument \"%d\" for \"--to-shard\" flag: please specify a valid shard ID using --to-shard and try again!", receiverShard)
+	}
+
+	return nil
+}
+
+// ValidateShard validates that a shardID is within the bounds of the shardCount
+func ValidShardID(shardID int, shardCount int) bool {
+	if shardID < 0 || shardID > (shardCount-1) {
```

Comment:
I think cobra can say that the shard ID flag should be a uint, so can at least get rid of this `< 0` check at library level and offload resp to the CLI's entry point flag

Reply:
yes, see cmd/subcommands/transfer.go

```golang
// Should be Uint32 instead of the second line
	cmdTransfer.Flags().Uint32Var(p *uint32, name string, value uint32, usage string)
	cmdTransfer.Flags().IntVar(&fromShardID, "from-shard", -1, "source shard id")
```

Reply:
I tried switching it over to uint32 but that leads to a bunch of incompatibilities with method signatures and methods that check regular ints (e.g. len(s)) etc.

It can obviously be resolved using type conversion, but not sure what you prefer.

The default arguments for both from-shard and to-shard are also -1 - do you know why that is? Does "-1" represent a special value for shardIDs somewhere? Or is it just simply a placeholder for a non-specified shardID?

Let me know what you think and I'll change the code accordingly.

Reply:
but shardID can not be in any context that needs len. Its fine to thread through a `uint32(originalNumber)` 

-1 was a quick sentiental value to signal bad input

Reply:
The way the shardID validation currently works in this PR (as well as the last one) is that it checks the length of `sharding.Structure(node)` and compares that to the supplied shardIDs, e.g:

```
s, err := sharding.Structure(node)
if err != nil {
	return err
}
err = validation.ValidShardIDs(fromShardID, toShardID, len(s))
if err != nil {
	return err
}
```

I'll go ahead and switch everything over to uint32. I'll set the default value to 0 since -1 can't be used for uint32.